src/hotspot/cpu/x86/stubGenerator_x86_64_dilithium.cpp

```diff
@@ -813,201 +1133,186 @@ static address generate_dilithiumDecomposePoly_avx512(StubGenerator *stubgen,
-
-  __ evpsubd(xmm16, k0, xmm16, xmm26, false, Assembler::AVX_512bit);
-  __ evpsubd(xmm17, k0, xmm17, xmm26, false, Assembler::AVX_512bit);
-  __ evpsubd(xmm18, k0, xmm18, xmm26, false, Assembler::AVX_512bit);
-  __ evpsubd(xmm19, k0, xmm19, xmm26, false, Assembler::AVX_512bit);
-  // r1 in xmm16
+  for (int i = 0; i < regCnt; i++) {
+    __ vpsubd(RPlus[i], RPlus[i], R0[i], vector_len);
+  }
+
   // r1 = (r1 | (-r1)) >> 31; // 0 if rplus - r0 == (dilithium_q - 1), -1 otherwise
-  __ evpsubd(xmm20, k0, zero, xmm16, false, Assembler::AVX_512bit);
-  __ evpsubd(xmm21, k0, zero, xmm17, false, Assembler::AVX_512bit);
-  __ evpsubd(xmm22, k0, zero, xmm18, false, Assembler::AVX_512bit);
-  __ evpsubd(xmm23, k0, zero, xmm19, false, Assembler::AVX_512bit);
-
-  __ evporq(xmm16, k0, xmm16, xmm20, false, Assembler::AVX_512bit);
-  __ evporq(xmm17, k0, xmm17, xmm21, false, Assembler::AVX_512bit);
-  __ evporq(xmm18, k0, xmm18, xmm22, false, Assembler::AVX_512bit);
-  __ evporq(xmm19, k0, xmm19, xmm23, false, Assembler::AVX_512bit);
-
-  __ evpsubd(xmm12, k0, zero, one, false, Assembler::AVX_512bit); // -1
-
-  __ evpsrad(xmm0, k0, xmm16, 31, false, Assembler::AVX_512bit);
-  __ evpsrad(xmm1, k0, xmm17, 31, false, Assembler::AVX_512bit);
-  __ evpsrad(xmm2, k0, xmm18, 31, false, Assembler::AVX_512bit);
-  __ evpsrad(xmm3, k0, xmm19, 31, false, Assembler::AVX_512bit);
-  // r1 in xmm0
-  // r0 += ~r1;
-  __ evpxorq(xmm20, k0, xmm0, xmm12, false, Assembler::AVX_512bit);
-  __ evpxorq(xmm21, k0, xmm1, xmm12, false, Assembler::AVX_512bit);
-  __ evpxorq(xmm22, k0, xmm2, xmm12, false, Assembler::AVX_512bit);
-  __ evpxorq(xmm23, k0, xmm3, xmm12, false, Assembler::AVX_512bit);
-
-  __ evpaddd(xmm8, k0, xmm8, xmm20, false, Assembler::AVX_512bit);
-  __ evpaddd(xmm9, k0, xmm9, xmm21, false, Assembler::AVX_512bit);
-  __ evpaddd(xmm10, k0, xmm10, xmm22, false, Assembler::AVX_512bit);
-  __ evpaddd(xmm11, k0, xmm11, xmm23, false, Assembler::AVX_512bit);
-  // r0 in xmm8
-  // r1 = r1 & quotient;
-  __ evpandd(xmm0, k0, xmm4, xmm0, false, Assembler::AVX_512bit);
-  __ evpandd(xmm1, k0, xmm5, xmm1, false, Assembler::AVX_512bit);
-  __ evpandd(xmm2, k0, xmm6, xmm2, false, Assembler::AVX_512bit);
-  __ evpandd(xmm3, k0, xmm7, xmm3, false, Assembler::AVX_512bit);
-  // r1 in xmm0
+  if (vector_len == Assembler::AVX_512bit) {
+    KRegister EqMsk[] = {k1, k2, k3, k4};
+    for (int i = 0; i < regCnt; i++) {
+      __ evpcmpeqd(EqMsk[i], k0, RPlus[i], qMinus1, vector_len);
+    }
+
+    // r0 += ~r1; // add -1 or keep as is, using EqMsk as filter
+    for (int i = 0; i < regCnt; i++) {
+      __ evpaddd(R0[i], EqMsk[i], R0[i], minusOne, true, vector_len);
+    }
+
+    // r1 in Quotient
+    // r1 = r1 & quotient; // copy 0 or keep as is, using EqMsk as filter
+    for (int i = 0; i < regCnt; i++) {
+      // FIXME: replace with void evmovdqul(Address dst, KRegister mask, XMMRegister src, bool merge, int vector_len);?
```

Comment:
Is the fixme a leftover?

Reply:
Yes. Removed. (I think I was considering merging this instruction with the storeXmm, but there really isnt a good way to do that)